rfd/0091-session-streaming.md

```diff
@@ -0,0 +1,238 @@
+last N events, this approach is not suitable here for several reasons:
+
+- There is not an obvious buffer size to use. Select too small of a buffer, and
+  you end up needing to start from the beginning anyway. Select too large of a
+  buffer and you end up in the same state we are in today 
+  (too much memory usage, high latency, etc)
+- The state of the terminal at any point in time is a function of all of the
+  events that have occurred up until that time, so you always need to play from
+  the beginning to accurately depict the session. This is true for both TTY 
+  sessions and desktop sessions.
+
+Here we propose the naive approach of rewinding by:
+
+1. restarting the stream from the beginning
+2. "fast forwarding" up to the desired rewind point by playing back events 
+   without inserting a timing delay
+3. resuming playback from this point at regular speed
+
+### `tsh`
+
+For `tsh` we can leverage the gRPC call that wraps `StreamSessionEvents`.
+`tsh` will perform playback in a background goroutine, and listen for keypresses
+to control playback (play, pause, skip forward, etc).
+
+`tsh` supports the following playback commands:
+
+- `ctrl+c` / `ctrl+d`: stop playing
+- `space`: play/pause
+- `left` / `down`: rewind
+- `right` / `up`: fast-forward
+
+This will be a relatively isolated change, as the `tsh` player code already
+ranges over a slice of events. The loop will need to range over a channel of
+events instead.
+
+### Web UI
+
+Today, the web UI performs an API call to 
+`/webapi/sites/:site/sessions/:sid/stream`. Unlike other APIs, this endpoint
+does not return JSON - it returns the text contents directly. This API call
+is synchronous and does not support streaming, it writes the entire session
+in a single response.
+
+In order to better support streaming, we will need a new endpoint with better
+support for a streaming API. The two options considered are server-sent events 
+(SSE) and websockets.
+
+While SSE is often a good fit for streaming applications, it suffers from two
+major limitations. First, most browsers implement a relatively small limit (6)
+on the number of SSE subscriptions (player tabs, in our case) that can be open 
+to a single site. Second, while SSE is a convenient mechansim for a server to 
+stream data to a client, we actually need a bidirectional stream so that the
+client can stream commands (play, pause, etc.) to the server.
+
+This leaves websockets as the best choice for web UI playback. In addition to
+supporting bidirectional streaming, websockets are already used throughout 
+Teleport and will not require additional server-side dependencies.
+
+To implement this, we can leverage the desktop session player for inspiration.
+We will add a new endpoint, `/webapi/sites/:site/sshplayback/:sid`, which 
+upgrades the connection to websockets and streams events to the browser while
+listening for playback commands. The data flowing over this websocket connection
+will be JSON-encoded.
```

Comment:
How big will the packets sent from the server to the client be?

We could probably utilise binary encoding on the websocket level for speed purposes and then have the Go handler parse binary to JSON for parsing the commands

Reply:
Yes, we do exactly this for desktop playback today. I'll clarify that here.